site/src/pages/WorkspacePage/useResourcesNav.test.tsx

```diff
@@ -0,0 +1,125 @@
+        />
+      ),
+    });
+    expect(result.current.selected?.id).toBe(MockWorkspaceResource.id);
+  });
+
+  it("selects the first resource if it has agents and selected resource is not find", async () => {
+    const resources: WorkspaceResource[] = [
+      MockWorkspaceResource,
+      {
+        ...MockWorkspaceResource,
+        agents: [],
+      },
+    ];
+    const { result } = renderHook(() => useResourcesNav(resources), {
+      wrapper: ({ children }) => (
+        <RouterProvider
+          router={createMemoryRouter([{ path: "/", element: children }], {
+            initialEntries: ["/?resources=not_found_resource_id"],
+          })}
+        />
+      ),
+    });
+    expect(result.current.selected?.id).toBe(MockWorkspaceResource.id);
+  });
+
+  it("selects the resource passed in the URL", () => {
+    const resources: WorkspaceResource[] = [
+      {
+        ...MockWorkspaceResource,
+        type: "docker_container",
+        name: "coder_python",
+      },
+      {
+        ...MockWorkspaceResource,
+        type: "docker_container",
+        name: "coder_java",
+      },
+      {
+        ...MockWorkspaceResource,
+        type: "docker_image",
+        name: "coder_image_python",
+        agents: [],
+      },
+    ];
+    const { result } = renderHook(() => useResourcesNav(resources), {
+      wrapper: ({ children }) => (
+        <RouterProvider
+          router={createMemoryRouter([{ path: "/", element: children }], {
+            initialEntries: [`/?resources=${resourceOptionId(resources[1])}`],
+          })}
+        />
+      ),
+    });
+    expect(result.current.selected?.id).toBe(resources[1].id);
+  });
+
+  it("selects a resource when resources are updated", () => {
+    const startedResources: WorkspaceResource[] = [
+      {
+        ...MockWorkspaceResource,
+        type: "docker_container",
+        name: "coder_python",
+      },
+      {
+        ...MockWorkspaceResource,
+        type: "docker_container",
+        name: "coder_java",
+      },
+      {
+        ...MockWorkspaceResource,
+        type: "docker_image",
+        name: "coder_image_python",
+        agents: [],
+      },
+    ];
+    const { result, rerender } = renderHook(
+      ({ resources }) => useResourcesNav(resources),
+      {
+        wrapper: ({ children }) => (
+          <RouterProvider
+            router={createMemoryRouter([{ path: "/", element: children }])}
+          />
+        ),
+        initialProps: { resources: startedResources },
+      },
+    );
+    expect(result.current.selected?.id).toBe(startedResources[0].id);
+
+    //  When a workspace is stopped, there is no resource with agents
+    const stoppedResources: WorkspaceResource[] = [
+      {
+        ...MockWorkspaceResource,
+        type: "docker_image",
+        name: "coder_image_python",
+        agents: [],
+      },
+    ];
+    rerender({ resources: stoppedResources });
+    expect(result.current.selected).toBe(undefined);
```

Comment:
Do you think it makes sense to add testing logic to check that the URL is syncing correctly on renders/re-renders, too?

Reply:
I think it is too much since we don't care too much about the URL but what the hook is returning. What use case could justify this test?

Reply:
It's more for making sure that the value we're getting back isn't getting out of sync with the URL. I don't think the current implementation has any real risks of that, but the way I see the hook, it does two things:

- Exposes the selection values
- Fires side effects to sync the selections with the URL

And we're only testing one of those things